README
**Nexa Prerequisites**
======================

    **Setting Up Cassandra**
    ------------------------

        **Prerequisites**

            Install Oracle Java Standard Edition 8 and Python 2.7 from your distro's package manager.

        **Installation**

            We are going to install all the necessary software under /opt directory which has sufficient disk space 
            (ideally 1 TB) and is mounted on an SSD.

            Download Apache Cassandra 3.11.6 from this official download link and unarchive directly under /opt.

        **Configuration**

            Edit the config file (/opt/apache-cassandra-3.11.6/conf/cassandra.yaml)

            Update the listen address appropriately,

            listen_address: localhost

            Also tweak the below settings appropriately per the instructions provided in the config file. 
            The below sample values are good for a 12 core machine

            concurrent_reads: 32 
            concurrent_writes: 96 
            concurrent_counter_writes: 32

            Also explicitly set the max threads configuration as below,

            native_transport_max_threads: 256

            To facilitate the highly concurrent application, make the following system configurations, it primarily 
            is meant to help Cassandra operate optimally.

            /etc/security/limits.conf
            * - memlock unlimited
            * - nofile 100000
            * - nproc 32768
            * - as unlimited

            /etc/sysctl.conf
            vm.max_map_count = 1048575

            Set the PATH environment variable as follows in your .bashrc file:

            export PATH=$PATH:/opt/apache-cassandra-3.11.6/bin

            Cassandra can be started using the following command

            $cassandra

            To stop Cassandra, we can simply get the processID and kill it.

            $ps -ef |grep cassandra
            $kill -9 <pid>

        **Database schema setup**

            Login to CQL shell (cqlsh) and execute the contents of the schema.cql file that was included in the release.

    **Setting Up Neo4j**
    --------------------

        **Installation**

            Download the official (3.5.20) release of Neo4j from here and unarchive the same under /opt.

            Note: Only Neo4j version "3.5.x" is officially supported. The packaged Neo4j driver does NOT support versions 4.x.x.

        **Configuration**

            Edit the config file at /opt/neo4j-community-3.5.20/conf/neo4j.conf and add the following at the bottom of the file

            dbms.connector.bolt.thread_pool_min_size=20
            dbms.connector.bolt.thread_pool_max_size=2048
            dbms.connector.bolt.thread_pool_keep_alive=5m

            Also uncomment and update the following entry appropriately,

            dbms.connector.default_listen_address=0.0.0.0

            Login to the Neo4j web console (http://localhost:7474/) and execute the following:

            CREATE CONSTRAINT ON (n:mnode) ASSERT n.v IS UNIQUE;
            
            You can verify if the constraints have been applied correctly by querying

            $ :schema
            And you should see the following output,

            Indexes
            ON :mnode(v) ONLINE  (for uniqueness constraint)

            Constraints
            ON ( mnode:mnode ) ASSERT mnode.v IS UNIQUE
            
            Set the PATH environment variable as follows in your .bashrc file:

            export PATH=$PATH:/opt/apache-cassandra-3.11.6/bin:/opt/neo4j-community-3.5.20/bin

            neo4j can be started and stopped using the following command/s

            $neo4j [start/stop/restart]

    **Other packages**
    ------------------

        Using the appropriate package manager for your Linux distribution install the below packages. 
        For example on Arch variants run,

        sudo pacman -S libsecp256k1
        sudo pacman -S leveldb


**Nexa Installation**
=====================

    Download the latest stable release from the downloads page and unzip contents under /opt/xoken

    The Nexa node sources configurable parameters from two different config files, as below:

    - node-config.yaml: Contains all the general Nexa node, Bitcoin P2P & listen endpoint related configurations.
    - arivi-config.yaml: Contains all the Arivi P2P related configurations (typically autogenerated)

    **Configure**
    -------------
        Place the node-config.yaml (provided in the release) in the same directory as the xoken-nexa executable. 
        Typically you will need to update the following parameters.

        - bitcoinNodeListenIP : the Nexa host's public IP that is used to interact with public Bitcoin mining nodes via BitconP2P.
        - bitcoinNodeListenPort : the system port that is used to connect with public Bitcoin mining nodes. 
        - endPointTLSListenIP : the host's public IP where the TLS endpoint is listening.
        - endPointTLSListenPort : the port being used by the TLS server. Default value of '9090' is recommended.
        - tlsCertificatePath : the TLS self-signed certificate file path, including the file-name.
        - tlsKeyfilePath : the TLS keyfile path, including the file-name.
        - tlsCertificateStorePath : Please set to any valid directory, currently the client certificate validation is disabled.
        - endPointHTTPSListenIP : listen IP for the HTTPS REST API endpoint.
        - endPointHTTPSListenPort : listen port for the HTTPS REST API.
        - endpointneo4jUsername : neo4j username
        - neo4jPassword : neo4j password
        - logLevel : Debug/Warn/Error , Debug level is very verbose and is not recommended unless you are troubleshooting.

        The rest are mostly tuning related parameters and meant for advanced users.

        The following command can be used to generate a TLS certificate and keyfile pair:

        openssl req -newkey rsa:4096 -x509 -sha256 -days 3650 -nodes -out example.crt -keyout example.key

    **Running Nexa**
    ----------------

        The Xoken Nexa server can be started as below, and it will run as a daemon process by defualt

        $ ./xoken-nexa

        And to stop the node you can do

        $ pkill xoken-nexa

    **Backup admin password & keys**
    --------------------------------

        Once the server has started successfully for the first time, open the nexa.log file and safely backup 
        the auto generated admin password, this is needed to provision other users and control user/node privileges.

        Open the newly generated arivi-config.yaml file and safely backup the application generated secret key. 
        This is essential to restore the unique identity of your node, since nodes maintain the reputation of their peers.

        e.g. secretKey: d670be193389c79349edd8ea4194925609f39cbeffc59b78fc1b4d431c0ebac

ReleaseNotes

Xoken Nexa Node
===============

v1.2.0     2020-12-18

- Query OP_RETURN outputs by protocol hex strings, supports multiple levels for sub-protocols & cursor based pagination
- An intuitive Analytics API (custom query language resembling Graph-QL) for obtaining insights into OP_RETURN protocol (nested) statistics;
  provides data analytics over time, blocks, protocols etc. supports counts & aggregation on size, fees, count etc.
- General improvements & fixes.


v1.1.2     2020-11-05

- Unconfirmed transactions now included in most API responses 
- CQL performance improvements with prepared statements
- XCQL driver defect fixes
- Performance improvements and minor memory leaks addressed
- Added max-buffer and max-threads config params for parallel tx processing
- Segmentation for transactions larger than 16 MB
- Custom implementation of a thread-safe hash-table
- Chain re-org sync stalling defect fix
- Stray unconfirmed Tx induced peer connection termination fix
- Schema order-by fix for epoch outputs table.


v1.1.1     2020-08-27

- Xcql Cassandra driver custom performance upgrades
- XCql full multiplexing support for requests/responses
- P2P streaming blocks, ingest batched transactions in chunks
- Improvements to concurrent syncing
- General bug fixes and code cleanup 


v1.1.0     2020-07-17

- HTTPS endpoint & REST API added, supports convenient JSON responses.
- Nexa now runs as a daemon process.
- User-based authorization added; users have roles (admin or regular user) with permissions that can be set 
- All calls to API authenticated with a session key that expires after its quota is used up or past an expiry time, whichever is earlier.
- API calls to authenticate and get the session key using credentials (username, password) added
- API calls to fetch and update user information, and add or delete users added.
- UTXO APIs added: get unspent outputs for given address[es]/scripthash[es]
- Spend status APIs added: get spend status and details for supplied transaction ID and output index
- Cursor-based pagination added for APIs that may return large amount of results; results include a nextCursor that can be supplied with subsequent requests for next page[s], along with a page size attribute
- Chain-related API calls added: get information about chain work and chain synchronization progress, and get chain headers given starting block height and number of blocks (acts as page size)
- JSON responses can be neatly formatted; this is configurable via 'pretty' Boolean switch for all API requests.
- Processing of unconfirmed transactions begins automatically after the chain sync is complete
- API to fetch all transaction IDs for a block, queried by block hash has been added
- Tx output spending information is now computed and stored in database
- Invalidation-less caching of Tx output values to speed up compilation of Tx input values and output spending information
- Cache key bits for invalidation-less cache can be configured from node-config.yaml to tune cache performance wherever invalidation-less caching is used
- Cassandra connection configuration tuned to better handle heavy loads
- Redesigned JSON request and response design, including RPC method parameters and response parameters to make information structure more intuitive and easier to comprehend
- Better resilience to Cassandra/Neo4j failures, Nexa will try to reconnect to DB[s] in case of failures
- Improvements to concurrent transaction processing, fixes that address possible race conditions
- A number of other parameters have been made configurable via node-config.yaml
- General bug fixes and code cleanup


v1.0.0-beta.1     2020-06-04

- Fix for chain sync slowdown
- Config file updates (files paths & DB credentials)


v1.0.0-beta     2020-05-26

- Direct TLS API endpoint support, running Arch is no longer necessary for clients with mere TLS capabilities. Clients have a choice of accessing Nexa either via AriviP2P protocol or TLS/Cbor
- Numerous performance improvements, syncs chain much faster.
- Split configuration parameters into two exclusive files focusing on specialized areas.
- Concurrent Tx processing, more efficient handling of out of sequence Tx processing i.e. until input dependencies are met.
- Tx broadcast API added
- Temporary blacklisting of nodes that aren't pointing to the canonical chain.
- More efficient Transpose Merkle Tree builder.
- Performance tuning parameters to tailor capabilities based on the system hardware resources.
- Invalidation-less Unconfirmed-Tx caching, to avoid triggering GetData for Tx INV messages.
- Bugfix to address stalled syncing on resource constrained hardware.
- General bug fixes and code cleanup.


v1.0.0-alpha    2020-02-05

- First version, released on an unsuspecting world.


node/app/Main.hs
{-# LANGUAGE ApplicativeDo #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE MultiWayIf #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE TemplateHaskell #-}
{-# OPTIONS_GHC -fno-warn-orphans #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE StandaloneDeriving #-}
{-# LANGUAGE UndecidableInstances #-}
{-# LANGUAGE BlockArguments #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE BangPatterns #-}

import Arivi.Crypto.Utils.PublicKey.Signature as ACUPS
import Arivi.Crypto.Utils.PublicKey.Utils
import Arivi.Crypto.Utils.Random
import Arivi.Env
import Arivi.Network
import Arivi.P2P
import qualified Arivi.P2P.Config as Config
import Arivi.P2P.Kademlia.Types
import Arivi.P2P.P2PEnv as PE hiding (option)
import Arivi.P2P.PubSub.Types
import Arivi.P2P.RPC.Types
import Arivi.P2P.ServiceRegistry
import Control.Arrow
import Control.Concurrent (threadDelay)
import Control.Concurrent
import qualified Control.Concurrent.Async as A (async, uninterruptibleCancel)
import Control.Concurrent.Async.Lifted as LA (async, race, wait, withAsync)
import Control.Concurrent.Event as EV
import Control.Concurrent.MSem as MS
import Control.Concurrent.MVar
import Control.Concurrent.QSem
import Control.Concurrent.STM.TQueue as TB
import Control.Concurrent.STM.TVar
import Control.Exception (throw)
import Control.Monad
import Control.Monad
import Control.Monad.Base
import Control.Monad.Catch
import Control.Monad.Except
import Control.Monad.Logger
import Control.Monad.Loops
import Control.Monad.Reader
import qualified Control.Monad.STM as CMS (atomically)
import Control.Monad.Trans.Control
import Control.Monad.Trans.Maybe
import Data.Aeson.Encoding (encodingToLazyByteString, fromEncoding)
import Data.Bits
import qualified Data.ByteString as B
import qualified Data.ByteString.Base16 as B16
import Data.ByteString.Base64 as B64
import Data.ByteString.Builder
import qualified Data.ByteString.Char8 as C
import qualified Data.ByteString.Lazy as L
import qualified Data.ByteString.Lazy.Char8 as CL
import Data.Char
import Data.Default
import Data.Default
import Data.Function
import Data.Functor.Identity
import qualified Data.HashTable as CHT
import qualified Data.HashTable.IO as H
import Data.IORef
import Data.Int
import Data.List
import Data.Map.Strict as M
import Data.Maybe
import Data.Maybe
import Data.Pool
import Data.Serialize as Serialize
import Data.Serialize as S
import Data.String.Conv
import Data.String.Conversions
import qualified Data.Text as DT
import qualified Data.Text as T
import qualified Data.Text.Encoding as DTE
import qualified Data.Text.Lazy as TL
import Data.Time.Calendar
import Data.Time.Clock
import Data.Time.Clock.POSIX
import Data.Typeable
import Data.Version
import Data.Word (Word32)
import Data.Word
import qualified Database.Bolt as BT
import qualified Database.XCQL.Protocol as Q
import Network.Simple.TCP
import Network.Socket
import Network.Xoken.Node.AriviService
import Network.Xoken.Node.Data
import Network.Xoken.Node.Data.ThreadSafeHashTable as TSH
import Network.Xoken.Node.Env
import Network.Xoken.Node.GraphDB
import Network.Xoken.Node.HTTP.Server
import Network.Xoken.Node.P2P.BlockSync
import Network.Xoken.Node.P2P.ChainSync
import Network.Xoken.Node.P2P.Common
import Network.Xoken.Node.P2P.PeerManager
import Network.Xoken.Node.P2P.Types
import Network.Xoken.Node.P2P.UnconfTxSync
import Network.Xoken.Node.TLSServer
import Options.Applicative
import Paths_xoken_node as P
import Prelude as P
import qualified Snap as Snap
import StmContainers.Map as SM
import System.Directory (doesDirectoryExist, doesFileExist)
import System.Environment (getArgs)
import System.Exit
import System.FilePath
import System.IO.Unsafe
import qualified System.Logger as LG
import qualified System.Logger.Class as LGC
import System.Posix.Daemon
import System.Random
import Text.Read (readMaybe)
import Xoken
import Xoken.Node
import Xoken.NodeConfig as NC

newtype AppM a =
    AppM (ReaderT (ServiceEnv) (IO) a)
    deriving (Functor, Applicative, Monad, MonadReader (ServiceEnv), MonadIO, MonadThrow, MonadCatch)

deriving instance MonadBase IO AppM

deriving instance MonadBaseControl IO AppM

instance HasBitcoinP2P AppM where
    getBitcoinP2P = asks (bitcoinP2PEnv . xokenNodeEnv)

instance HasDatabaseHandles AppM where
    getDB = asks (dbHandles . xokenNodeEnv)

instance HasAllegoryEnv AppM where
    getAllegory = asks (allegoryEnv . xokenNodeEnv)

instance HasLogger AppM where
    getLogger = asks (loggerEnv . xokenNodeEnv)

runAppM :: ServiceEnv -> AppM a -> IO a
runAppM env (AppM app) = runReaderT app env

data ConfigException
    = ConfigParseException
    | RandomSecretKeyException
    deriving (Eq, Ord, Show)

instance Exception ConfigException

type HashTable k v = H.BasicHashTable k v

defaultConfig :: IO ()
defaultConfig = do
    (sk, _) <- ACUPS.generateKeyPair
    let bootstrapPeer =
            Peer
                ((fst . B16.decode)
                     "a07b8847dc19d77f8ef966ba5a954dac2270779fb028b77829f8ba551fd2f7ab0c73441456b402792c731d8d39c116cb1b4eb3a18a98f4b099a5f9bdffee965c")
                (NodeEndPoint "51.89.40.95" 5678 5678)
    let config =
            Config.Config 5678 5678 sk [bootstrapPeer] (generateNodeId sk) "127.0.0.1" (T.pack "./arivi.log") 20 5 3
    Config.makeConfig config "./arivi-config.yaml"

makeGraphDBResPool :: T.Text -> T.Text -> IO (ServerState)
makeGraphDBResPool uname pwd = do
    let gdbConfig = def {BT.user = uname, BT.password = pwd}
    gdbState <- constructState gdbConfig
    a <- withResource (pool gdbState) (`BT.run` queryGraphDBVersion)
    putStrLn $ "Connected to Neo4j database, version " ++ show (a !! 0)
    return gdbState

initXCql :: NC.NodeConfig -> IO (XCqlClientState)
initXCql nodeConf = do
    let hints = defaultHints {addrFlags = [AI_NUMERICHOST, AI_NUMERICSERV], addrSocketType = Stream}
        startCql :: Q.Request k () ()
        startCql = Q.RqStartup $ Q.Startup Q.Cqlv300 (Q.algorithm Q.noCompression) --(Q.CqlVersion "3.4.4") Q.None
    mapM
        (\cn -> do
             (addr:_) <- getAddrInfo (Just hints) (Just "127.0.0.1") (Just "9042")
             s <- socket (addrFamily addr) (addrSocketType addr) (addrProtocol addr)
             Network.Socket.connect s (addrAddress addr)
             connHandshake s startCql
             t <- TSH.new 1
             l <- newMVar (1 :: Int16)
             sk <- newIORef $ Just s
             let xcqlc = XCQLConnection t l sk
             A.async (readResponse xcqlc)
             return xcqlc)
        [1 .. (maxConnectionsXCql nodeConf)]

runThreads :: Config.Config -> NC.NodeConfig -> BitcoinP2P -> XCqlClientState -> LG.Logger -> [FilePath] -> IO ()
runThreads config nodeConf bp2p conn lg certPaths = do
    gdbState <- makeGraphDBResPool (neo4jUsername nodeConf) (neo4jPassword nodeConf)
    let dbh = DatabaseHandles gdbState conn
    let allegoryEnv = AllegoryEnv $ allegoryVendorSecretKey nodeConf
    let xknEnv = XokenNodeEnv bp2p dbh lg allegoryEnv
    let serviceEnv = ServiceEnv xknEnv
    epHandler <- newTLSEndpointServiceHandler
    -- start TLS endpoint
    async $ startTLSEndpoint epHandler (endPointTLSListenIP nodeConf) (endPointTLSListenPort nodeConf) certPaths
    -- start HTTP endpoint
    let snapConfig =
            Snap.defaultConfig & Snap.setSSLBind (DTE.encodeUtf8 $ DT.pack $ endPointHTTPSListenIP nodeConf) &
            Snap.setSSLPort (fromEnum $ endPointHTTPSListenPort nodeConf) &
            Snap.setSSLKey (certPaths !! 1) &
            Snap.setSSLCert (head certPaths) &
            Snap.setSSLChainCert False
    async $ Snap.serveSnaplet snapConfig (appInit xknEnv)
    withResource (pool $ graphDB dbh) (`BT.run` initAllegoryRoot genesisTx)
    -- run main workers
    runAppM
        serviceEnv
        (do bp2pEnv <- getBitcoinP2P
            withAsync runEpochSwitcher $ \_ -> do
                withAsync setupSeedPeerConnection $ \_ -> do
                    withAsync runEgressChainSync $ \_ -> do
                        withAsync runBlockCacheQueue $ \_ -> do
                            withAsync (handleNewConnectionRequest epHandler) $ \_ -> do
                                withAsync runPeerSync $ \_ -> do
                                    withAsync runSyncStatusChecker $ \_ -> do
                                        withAsync runWatchDog $ \z -> do
                                            _ <- LA.wait z
                                            return ())
    liftIO $ destroyAllResources $ pool gdbState
    liftIO $ putStrLn $ "node recovering from fatal DB connection failure!"
    return ()

runSyncStatusChecker :: (HasXokenNodeEnv env m, HasLogger m, MonadIO m) => m ()
runSyncStatusChecker = do
    lg <- getLogger
    bp2pEnv <- getBitcoinP2P
    conn <- xCqlClientState <$> getDB
    liftIO $ threadDelay (300 * 1000000)
    forever $ do
        isSynced <- checkBlocksFullySynced conn
        liftIO $ CMS.atomically $ writeTVar (indexUnconfirmedTx bp2pEnv) isSynced
        liftIO $ threadDelay (10 * 1000000)

runWatchDog :: (HasXokenNodeEnv env m, MonadIO m) => m ()
runWatchDog = do
    dbe <- getDB
    lg <- getLogger
    bp2pEnv <- getBitcoinP2P
    let conn = xCqlClientState (dbe)
    LG.debug lg $ LG.msg $ LG.val "Starting watchdog"
    continue <- liftIO $ newIORef True
    whileM_ (liftIO $ readIORef continue) $ do
        liftIO $ threadDelay (120 * 1000000)
        fa <- liftIO $ readTVarIO (txProcFailAttempts bp2pEnv)
        if fa > 5
            then do
                LG.err lg $ LG.msg $ LG.val "Error: exceeded Tx proc fail attempt threshold, watchdog raise alert "
                liftIO $ writeIORef continue False
            else do
                tm <- liftIO $ getCurrentTime
                let ttime = (floor $ utcTimeToPOSIXSeconds tm) :: Int64
                    str = "insert INTO xoken.transactions ( tx_id, block_info, tx_serialized ) values (?, ?, ?)"
                    qstr = str :: Q.QueryString Q.W (T.Text, ((T.Text, Int32), Int32), Q.Blob) ()
                    par =
                        getSimpleQueryParam ("watchdog-last-check", ((T.pack $ show ttime, 0), 0), Q.Blob $ CL.pack "")
                ores <-
                    LA.race
                        (liftIO $ threadDelay (3000000)) -- worst case of 3 secs
                        (liftIO $ try $ write conn (Q.RqQuery $ Q.Query qstr par))
                case ores of
                    Right (eth) -> do
                        case eth of
                            Right _ -> return ()
                            Left (SomeException e) -> do
                                LG.err lg $ LG.msg ("Error: unable to insert, watchdog raise alert " ++ show e)
                                liftIO $ writeIORef continue False
                    Left () -> do
                        LG.err lg $ LG.msg $ LG.val "Error: insert timed-out, watchdog raise alert "
                        liftIO $ writeIORef continue False

runNode :: Config.Config -> NC.NodeConfig -> XCqlClientState -> BitcoinP2P -> [FilePath] -> IO ()
runNode config nodeConf conn bp2p certPaths = do
    lg <-
        LG.new
            (LG.setOutput
                 (LG.Path $ T.unpack $ NC.logFileName nodeConf)
                 (LG.setLogLevel (logLevel nodeConf) LG.defSettings))
    runThreads config nodeConf bp2p conn lg certPaths

data Config =
    Config
        { configNetwork :: !Network
        , configDebug :: !Bool
        , configUnconfirmedTx :: !Bool
        }

defPort :: Int
defPort = 3000

defNetwork :: Network
defNetwork = bsvTest

netNames :: String
netNames = intercalate "|" (Data.List.map getNetworkName allNets)

defaultAdminUser :: XCqlClientState -> IO ()
defaultAdminUser conn = do
    let qstr =
            " SELECT password from xoken.user_permission where username = 'admin' " :: Q.QueryString Q.R () (Identity T.Text)
        p = getSimpleQueryParam ()
    op <- query conn (Q.RqQuery $ Q.Query qstr p)
    if length op == 1
        then return ()
        else do
            tm <- liftIO $ getCurrentTime
            usr <-
                addNewUser
                    conn
                    "admin"
                    "default"
                    "user"
                    ""
                    (Just ["admin"])
                    (Just 100000000)
                    (Just (addUTCTime (nominalDay * 365) tm))
            putStrLn $ "******************************************************************* "
            putStrLn $ "  Creating default Admin user!"
            putStrLn $ "  Please note down admin password NOW, will not be shown again."
            putStrLn $ "  Password : " ++ (aurPassword $ fromJust usr)
            putStrLn $ "******************************************************************* "

defBitcoinP2P :: NodeConfig -> Int64 -> IO (BitcoinP2P)
defBitcoinP2P nodeCnf ept = do
    g <- newTVarIO M.empty
    bp <- newTVarIO M.empty
    mv <- newMVar True
    hl <- newMVar True
    st <- TSH.new 1
    tl <- TSH.new 4
    ep <- newTVarIO False
    epts <- newTVarIO ept
    tc <- TSH.new 16
    -- vc <- TSH.new 8 -- TSH.new 16
    rpf <- newEmptyMVar
    rpc <- newTVarIO 0
    mq <- TSH.new 4
    ts <- TSH.new 4
    tbt <- MS.new $ maxTMTBuilderThreads nodeCnf
    iut <- newTVarIO False
    udc <- H.new
    tpfa <- newTVarIO 0
    bsb <- newTVarIO Nothing
    pi <- TSH.new 32
    return $ BitcoinP2P nodeCnf g bp mv hl st tl ep epts tc (rpf, rpc) mq ts tbt iut udc tpfa bsb pi

initNexa :: IO ()
initNexa = do
    putStrLn $ "Starting Xoken Nexa"
    b <- doesFileExist "arivi-config.yaml"
    unless b defaultConfig
    cnf <- Config.readConfig "arivi-config.yaml"
    nodeCnf <- NC.readConfig "node-config.yaml"
    !conn <- initXCql nodeCnf
    defaultAdminUser conn
    tm <- getCurrentTime
    bp2p <- defBitcoinP2P nodeCnf (floor $ utcTimeToPOSIXSeconds tm)
    let certFP = tlsCertificatePath nodeCnf
        keyFP = tlsKeyfilePath nodeCnf
        csrFP = tlsCertificateStorePath nodeCnf
    cfp <- doesFileExist certFP
    kfp <- doesFileExist keyFP
    csfp <- doesDirectoryExist csrFP
    unless (cfp && kfp && csfp) $ P.error "Error: missing TLS certificate or keyfile"
    -- launch node --
    runNode cnf nodeCnf conn bp2p [certFP, keyFP, csrFP]

relaunch :: IO ()
relaunch =
    forever $ do
        let pid = "/tmp/nexa.pid.1"
        running <- isRunning pid
        if running
            then threadDelay (30 * 1000000)
            else do
                runDetached (Just pid) (ToFile "nexa.log") initNexa
                threadDelay (5000000)

main :: IO ()
main = do
    let pid = "/tmp/nexa.pid.0"
    runDetached (Just pid) (ToFile "nexa.log") relaunch
